### ISSUE TITLE
Task<Option<T>>.BindT throws ValueIsNoneException when in None state

### ISSUE THREAD

**Issue by bender2k14** [louthy/language-ext#381]
The following test fails in the current versions of both the `master` and `async-refactor` branches.  In both cases, `BindT` throws a `ValueIsNoneException`.

```
[Fact]
public async Task TaskOptionBindT_InitialOptionInNoneState_NoExceptionThrown()
{
    Option<Unit> x = None;
    var none = x.AsTask();
    var task = none.BindT(_ => none);
    
    await task;
}
```

**Comment by louthy** [comment]
That's almost certainly an artefact of the default `Fail` function for `MTask`.  With `BindT` the outer monad is in control of the bind operation, if the inner monad is in a failed state then it will fail too.  It's propagating the failure exception up the chain.  It's best to see the transformer types as one type rather than two nested ones, but with the rules of the outer monad being law.  The way `Task` deals with failure is an exception, so that's what you see.

I am super short of time at the moment, so I'll take a look when I can, but there will just be some combinations of types used in the generic transformer system that don't play well with each other (and that means manually unwrapping to inject the behaviour you want).

**Comment by bender2k14** [comment]
> It's best to see the transformer types as one type rather than two nested ones...

Indeed.  In fact, the workaround we found before creating this failing test was to call `.ToOptionAsync()` to transform `Task<Option<T>>` to `OptionAsync<T>` (in both places) and then replace `BindT` with `Bind`.

This could be the fix...to use `OptionAsync`.  Some possibilities of achieving this include

1. completely remove `BindT` and force the user to call `.ToOptionAsync()` themselves or
2. change the return type of `BindT` to `OptionAsync` and call `.ToOptionAsync()` internally.

> I am super short of time at the moment...

So no worries about not quickly coming up with a fix since a nice workaround exists.

**Comment by louthy** [comment]
This issue is fixed in `v2.2.17-beta`
